client/packages/host/src/Admin/WebAppLogFileModal.tsx

```diff
@@ -0,0 +1,145 @@
+import React, { useEffect, useState } from 'react';
+import { useTranslation } from '@common/intl';
+import {
+  BasicSpinner,
+  DialogButton,
+  DropdownMenu,
+  DropdownMenuItem,
+  Typography,
+  Box,
+  useDialog,
+  useNotification,
+  useWebClient,
+} from '@openmsupply-client/common';
+import { useLog } from '@openmsupply-client/system';
+import { LogTextDisplay } from './LogTextDisplay';
+
+export const LogDisplay = ({
+  fileName,
+  setLogContent,
+}: {
+  fileName: string;
+  setLogContent: (content: string[]) => void;
+}) => {
+  const { mutateAsync, data, isLoading } =
+    useLog.document.logContentsByFileName();
+
+  useEffect(() => {
+    if (!!fileName) {
+      mutateAsync(fileName);
+    }
+  }, [fileName]);
+
+  useEffect(() => {
+    if (!!data?.fileContent) {
+      setLogContent(data.fileContent);
+    }
+  }, [data]);
+
+  if (isLoading) {
+    return <BasicSpinner />;
+  }
+
+  return !!data?.fileContent ? (
+    <Box paddingTop={2} maxHeight={400}>
+      <LogTextDisplay logText={data?.fileContent} />
+    </Box>
+  ) : null;
+};
+
+export const WebAppLogFileModal = ({
+  isOpen,
+  onClose,
+}: {
+  isOpen: boolean;
+  onClose: () => void;
+}) => {
+  const t = useTranslation();
+  const { success, warning } = useNotification();
+  const [logToRender, setLogToRender] = useState('');
+  const [logContent, setLogContent] = useState<string[]>([]);
+  const { Modal } = useDialog({ isOpen });
+  const { saveFile } = useWebClient();
+  const [isSaving, setIsSaving] = useState(false);
+  const noLog = logContent.length === 0;
+
+  const { data, isError, isLoading } = useLog.document.listFileNames();
+
+  const saveLog = async () => {
+    if (noLog) {
+      warning(t('message.nothing-to-save'))();
+    } else if (isSaving) {
+      warning(t('message.already-saving'))();
+    } else {
+      setIsSaving(true);
+      saveFile({
+        content: logContent.toString(),
+        filename: logToRender,
+      });
+      setIsSaving(false);
+    }
+  };
+
+  const copyToClipboard = () => {
+    navigator.clipboard.writeText(logContent.toString()).then(() => {
+      success(t('message.copy-success'))();
+    });
+  };
+
+  if (isError) {
```

Comment:
It does throw an error if the folder/file doesn't exist.. you'll probably have to delete your logs and log to console only to see this error....
![Screenshot 2023-12-01 at 01 13 38](https://github.com/msupply-foundation/open-msupply/assets/61820074/72bebae9-c8a6-4903-adaf-70cb5f3c44f7)

Could be better handled in backend by returning structured errors... can be moved to another issue...

Reply:
Aha. I see - GqlContext is parsing the response and throwing an Error if there's an `error` returned in the valid response. that's good.

For my testing I just deleted my logs, but not the log folder itself! If you do that, then there's no error, and the fileNames array is empty - which I think should also get caught by this check.

Reply:
I also want to check for if data has been returned to prevent this error modal rendering while data is yet undefined, like so:

```  if (isError || (data && (data?.fileNames || []).length === 0)) {
...
```

Reply:
In this part: `(data && (data?.fileNames || []).length === 0)` the check for `data` is redundant.

`data?.filenames` will give undefined if `!data` and so the `|| []` is applied, and the length is `0`

which means that `(data?.fileNames || []).length === 0` does the same job. though maybe not as clear to read?

Reply:
Ah I might be thinking about this the wrong way -

I was thinking of the case above where you mentioned no error would be returned if there was a log folder but no log files:

```
{
    "data": {
        "logFileNames": {
            "__typename": "LogNode",
            "fileContent": null,
            "fileNames": []
        }
    }
}
```

In this case data is not undefined, but no logs are being returned.

Maybe this would result in alternative error handling or messaging?